### ISSUE TITLE
Add fix for duplicates caused by consume exiting too early

### ISSUE THREAD

**Issue by colmsnowplow** [snowplow-devops/kinsumer#11]
TestSplit results in duplicates sometimes. It's difficult to reproduce this issue directly in that integration test, but we consistently hit it over the course of ~7 runs of the test in CI. On local, when resources are constrained it occurs more often.

Deep dive investigation and tests to isolate different related behaviours (see #10) reveal that it occurs in TestSplit, when shards merge, because the shard count changes slowly over time, and there is a chance that for a period of time, different clients are operating upon the same shards. 

This alone doesn't produce duplicates, however when a k.stop() request is sent, [the consume function exits immediately](https://github.com/snowplow-devops/kinsumer/blob/7e157f4168a121540344312a5bf9b6a734eccb83/shard_consumer.go#L170), and pauses all checkpointing.

When this stop request happens after a record has already been pushed to `k.records` (and therefore out to the user), but before its `checkpointer.update()` function is called, that record is processed again by the next client to take ownership of the shard.

In fact, even when the shard doesn't change owner, the same problem exists if we stop and restart consuming the shard between pushing to` k.records` and calling `checkpointer.update()`.  Both of these scenarios can be isolated with (relatively convoluted) unit tests.

We might expect that [the deferred call to `release()` on exit mitigates this, but no so - I'm not sure if this is because it evaluates the checkpointer too early, or because a call to `update()` doesn't update the sequence number on time.

The solution to mitigate this issue is to amend the behaviour when a stop request comes in. We break from both the loop which executes commits/shard refreshes, and the loop which iterates the shard and returns records. Then, before returning, we start a new commit loop which commits as soon as we recognise a new sequence number, and exits only after a timeout.

This may introduce some latency when shard ownership changes - however such shard ownership changes are hopefully rare enough for this impact to be tolerable. This latency impact can also be managed by correlating the timeout to the now configurable maxAgeForClientRecord.

**Comment by colmsnowplow** [comment]
Resolved in https://github.com/snowplow-devops/kinsumer/pull/13
